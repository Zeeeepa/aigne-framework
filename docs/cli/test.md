# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
