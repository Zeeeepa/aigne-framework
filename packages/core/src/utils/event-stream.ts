import { createParser } from "eventsource-parser";
import { produce } from "immer";
import {
  type AgentResponseChunk,
  type AgentResponseProgress,
  type AgentResponseStream,
  isAgentResponseDelta,
  isAgentResponseProgress,
  type Message,
} from "../agents/agent.js";
import type { Context, ContextEventMap } from "../aigne/context.js";
import { tryOrThrow } from "./type-utils.js";
import type { Listener } from "./typed-event-emitter.js";

export class EventStreamParser<T> extends TransformStream<string, T | Error> {
  constructor() {
    let parser: ReturnType<typeof createParser> | undefined;

    super({
      start(controller) {
        parser = createParser({
          onEvent: (event) => {
            const json = tryOrThrow(
              () => JSON.parse(event.data) as T | { message: string },
              (e) => {
                controller.enqueue(
                  new Error(`Parse response chunk json error: ${e.message} ${event.data}`),
                );
              },
            );
            if (json) {
              switch (event.event) {
                case "error":
                  controller.enqueue(new Error((json as { message: string }).message));
                  break;
                default: {
                  if (!event.event) controller.enqueue(json as T);
                  else console.warn(`Unknown event type: ${event.event}`, event.data);
                }
              }
            }
          },
        });
      },
      transform(chunk) {
        parser?.feed(chunk);
      },
    });
  }
}

export class AgentResponseStreamParser<O extends Message> extends TransformStream<
  AgentResponseChunk<O> | Error,
  AgentResponseChunk<O>
> {
  private json: O = {} as O;

  constructor() {
    super({
      transform: (chunk, controller) => {
        if (chunk instanceof Error) {
          controller.error(chunk);
          controller.terminate();
          return;
        }

        if (isAgentResponseDelta(chunk)) {
          this.json = produce(this.json, (draft) => {
            if (chunk.delta.json) Object.assign(draft, chunk.delta.json);

            if (chunk.delta.text) {
              for (const [key, text] of Object.entries(chunk.delta.text)) {
                const original = draft[key] as string | undefined;
                const t = (original || "") + (text || "");
                if (t) Object.assign(draft, { [key]: t });
              }
            }
          });

          controller.enqueue({
            ...chunk,
            delta: {
              ...chunk.delta,
              json: this.json,
            },
          });
        } else if (isAgentResponseProgress(chunk)) {
          if (chunk.progress.event === "agentFailed") {
            const { name, message } = chunk.progress.error;
            chunk.progress.error = new Error(message);
            chunk.progress.error.name = name;
          }
          controller.enqueue(chunk);
        }
      },
    });
  }
}

export class AgentResponseStreamSSE<O extends Message> extends ReadableStream<string> {
  constructor(stream: AgentResponseStream<O>) {
    let reader: ReadableStreamDefaultReader<AgentResponseChunk<O>> | undefined;

    super({
      async pull(controller) {
        reader ??= stream.getReader();
        try {
          const { value, done } = await reader.read();
          if (done) {
            controller.close();
            return;
          }

          if (isAgentResponseProgress(value)) {
            if (value.progress.event === "agentFailed") {
              value.progress.error = {
                name: value.progress.error.name,
                message: value.progress.error.message,
              };
            }
          }

          controller.enqueue(`data: ${JSON.stringify(value)}\n\n`);
        } catch (error) {
          controller.enqueue(
            `event: error\ndata: ${JSON.stringify({ message: error.message })}\n\n`,
          );
          controller.close();
        }
      },
    });
  }
}

export class AgentResponseProgressStream extends ReadableStream<AgentResponseProgress> {
  constructor(context: Context) {
    super({
      async start(controller) {
        const writeEvent = (
          eventName: keyof ContextEventMap,
          event: ContextEventMap[typeof eventName][0],
        ) => {
          const progress = {
            ...event,
            event: eventName,
            agent: { name: event.agent.name },
          } as AgentResponseProgress["progress"];

          controller.enqueue({ progress });
        };

        const close = () => {
          context.off("agentStarted", onAgentStarted);
          context.off("agentSucceed", onAgentSucceed);
          context.off("agentFailed", onAgentFailed);
          controller.close();
        };

        const onAgentStarted: Listener<"agentStarted", ContextEventMap> = async (event) => {
          const taskTitle = await event.agent.renderTaskTitle(event.input);

          writeEvent("agentStarted", { ...event, taskTitle });
        };
        const onAgentSucceed: Listener<"agentSucceed", ContextEventMap> = (event) => {
          writeEvent("agentSucceed", event);
          if (event.contextId === context.id) {
            close();
          }
        };
        const onAgentFailed: Listener<"agentFailed", ContextEventMap> = (event) => {
          writeEvent("agentFailed", event);
          if (event.contextId === context.id) {
            close();
          }
        };

        context.on("agentStarted", onAgentStarted);
        context.on("agentSucceed", onAgentSucceed);
        context.on("agentFailed", onAgentFailed);
      },
    });
  }
}
