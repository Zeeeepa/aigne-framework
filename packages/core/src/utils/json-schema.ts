import type { ZodType, z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import type { Message } from "../agents/agent.js";
import { logger } from "./logger.js";

export function outputSchemaToResponseFormatSchema(
  agentOutput: ZodType<Message>,
): Record<string, unknown> {
  return setAdditionPropertiesDeep(zodToJsonSchema(agentOutput), false);
}

function setAdditionPropertiesDeep<T>(schema: T, additionalProperties: boolean): T {
  if (Array.isArray(schema)) {
    return schema.map((s) => setAdditionPropertiesDeep(s, additionalProperties)) as T;
  }
  if (schema !== null && typeof schema === "object" && !Array.isArray(schema)) {
    return Object.entries(schema).reduce(
      (acc, [key, value]) => {
        acc[key] = setAdditionPropertiesDeep(value, additionalProperties);
        if (acc.type === "object") {
          acc.additionalProperties = additionalProperties;
        }
        return acc;
      },
      {} as Record<string, unknown>,
    ) as T;
  }
  return schema;
}

export function parseJSON(json: string) {
  try {
    return JSON.parse(json);
  } catch (error) {
    logger.error("Failed to parse JSON", { json, error });
    throw error;
  }
}

/**
 * Ensure that the union array has at least 1 item.
 * NOTE: the zod union requires at least 2 items (just type definition, not runtime behavior)
 * so we need to ensure that the union has at least 1 item.
 * @param union - The union array
 * @returns The union array with at least 1 item (but the type is at least 2 items for z.union)
 */
export function ensureZodUnionArray<T extends z.ZodType>(union: T[]): [T, T, ...T[]] {
  if (!(union.length >= 1)) {
    throw new Error(`Union must have at least 1 item, but got ${union.length}`);
  }
  return union as [T, T, ...T[]];
}
