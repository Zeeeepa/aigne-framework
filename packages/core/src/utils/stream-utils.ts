import equal from "fast-deep-equal";
import {
  type AgentProcessAsyncGenerator,
  type AgentResponseChunk,
  type AgentResponseStream,
  isAgentResponseDelta,
  isEmptyChunk,
  type Message,
} from "../agents/agent.js";
import { isNonNullable, isRecord, omitBy, type PromiseOrValue } from "./type-utils.js";
import "./stream-polyfill.js";
import type { ReadableStreamDefaultReadResult } from "bun";

export function objectToAgentResponseStream<T extends Message>(json: T): AgentResponseStream<T> {
  if (!isRecord(json)) {
    throw new Error(
      `expect to return a record type such as {result: ...}, but got (${typeof json}): ${json}`,
    );
  }

  return new ReadableStream({
    pull(controller) {
      controller.enqueue({ delta: { json } });
      controller.close();
    },
  });
}

export function mergeAgentResponseChunk<T extends Message>(
  output: T,
  chunk: AgentResponseChunk<T>,
) {
  if (isAgentResponseDelta(chunk)) {
    if (chunk.delta.text) {
      for (const [key, text] of Object.entries(chunk.delta.text)) {
        const original = output[key] as string | undefined;
        const t = (original || "") + (text || "");
        if (t) Object.assign(output, { [key]: t });
      }
    }

    if (chunk.delta.json) {
      Object.assign(
        output,
        omitBy(chunk.delta.json, (v) => v === undefined),
      );
    }
  }

  return output;
}

export async function agentResponseStreamToObject<T extends Message>(
  stream: AgentResponseStream<T> | AgentProcessAsyncGenerator<T>,
): Promise<T> {
  const json: T = {} as T;

  if (stream instanceof ReadableStream) {
    for await (const value of stream) {
      mergeAgentResponseChunk(json, value);
    }
  } else {
    for (;;) {
      const chunk = await stream.next();
      if (chunk.value) {
        if (chunk.done) {
          Object.assign(json, chunk.value);
        } else {
          mergeAgentResponseChunk(json, chunk.value);
        }
      }
      if (chunk.done) break;
    }
  }

  return json;
}

export function asyncGeneratorToReadableStream<T extends Message>(
  generator: AgentProcessAsyncGenerator<T>,
): AgentResponseStream<T> {
  return new ReadableStream({
    async pull(controller) {
      try {
        const chunk = await generator.next();
        if (chunk.value) {
          if (chunk.done) {
            controller.enqueue({ delta: { json: chunk.value } });
          } else {
            controller.enqueue(chunk.value);
          }
        }

        if (chunk.done) {
          controller.close();
        }
      } catch (error) {
        controller.error(error);
      }
    },
  });
}

export function onAgentResponseStreamEnd<T extends Message>(
  stream: AgentResponseStream<T>,
  options?: {
    onChunk?: (
      chunk: AgentResponseChunk<T>,
    ) => PromiseOrValue<AgentResponseChunk<T> | undefined | void>;
    onResult?: (result: T) => PromiseOrValue<Partial<T> | undefined | void>;
    onError?: (error: Error) => PromiseOrValue<Error>;
  },
): AgentResponseStream<T> {
  const json: T = {} as T;
  const reader = stream.getReader();

  return new ReadableStream({
    async pull(controller) {
      try {
        while (true) {
          const { value, done } = await reader.read();

          if (done) {
            const result = (await options?.onResult?.(json)) ?? json;

            if (result && !equal(result, json)) {
              let chunk: AgentResponseChunk<T> = { delta: { json: result } };
              chunk = (await options?.onChunk?.(chunk)) ?? chunk;
              controller.enqueue(chunk);
            }

            controller.close();

            return;
          }

          mergeAgentResponseChunk(json, value);

          const chunk = (await options?.onChunk?.(value)) ?? value;
          if (!isEmptyChunk(chunk)) {
            controller.enqueue(chunk);
            break;
          }
        }
      } catch (error) {
        controller.error((await options?.onError?.(error)) ?? error);
      }
    },
  });
}

export function isAsyncGenerator<T extends AsyncGenerator>(
  value: AsyncGenerator | unknown,
): value is T {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}

export async function* arrayToAgentProcessAsyncGenerator<T extends Message>(
  chunks: (AgentResponseChunk<T> | Error)[],
  result?: Partial<T>,
): AgentProcessAsyncGenerator<T> {
  for (const chunk of chunks) {
    if (chunk instanceof Error) throw chunk;

    yield chunk;
  }
  if (result !== undefined) return result;
}

export function arrayToReadableStream<T>(chunks: (T | Error)[]): ReadableStream<T> {
  const list = [...chunks];

  return new ReadableStream({
    pull(controller) {
      const item = list.shift();
      if (!item) {
        controller.close();
        return;
      }

      if (item instanceof Error) {
        controller.error(item);
        return;
      }

      controller.enqueue(item);
    },
  });
}

export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options: { catchError: true },
): Promise<(T | Error)[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<T[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<(T | Error)[]> {
  const result: T[] = [];
  try {
    for await (const value of stream) {
      result.push(value);
    }
  } catch (error) {
    if (!options?.catchError) throw error;

    result.push(error);
  }
  return result;
}

export function stringToAgentResponseStream(
  str: string,
  key: "text" | string = "text",
): AgentResponseStream<Message> {
  const segmenter = new Intl.Segmenter(undefined, { granularity: "word" });
  const segments = segmenter.segment(str);

  return arrayToReadableStream(
    Array.from(segments).map((segment) => ({ delta: { text: { [key]: segment.segment } } })),
  );
}

export function toReadableStream(stream: NodeJS.ReadStream) {
  return new ReadableStream<Uint8Array>({
    start(controller) {
      const onData = (chunk: Buffer) => {
        controller.enqueue(new Uint8Array(chunk));
      };

      const onEnd = () => {
        cleanup();
        controller.close();
      };

      const onError = (err: Error) => {
        cleanup();
        controller.error(err);
      };

      function cleanup() {
        stream.off("data", onData);
        stream.off("end", onEnd);
        stream.off("error", onError);
      }

      stream.on("data", onData);
      stream.on("end", onEnd);
      stream.on("error", onError);
    },
  });
}

export async function readAllString(stream: NodeJS.ReadStream | ReadableStream): Promise<string> {
  return (
    await readableStreamToArray(
      (stream instanceof ReadableStream ? stream : toReadableStream(stream)).pipeThrough(
        new TextDecoderStream(),
      ),
    )
  ).join("");
}

export function mergeReadableStreams<T1, T2>(
  s1: ReadableStream<T1>,
  s2: ReadableStream<T2>,
): ReadableStream<T1 | T2>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any> {
  type Reader = {
    reader: ReadableStreamDefaultReader;
    reading?: Promise<{
      result: ReadableStreamDefaultReadResult<any>;
      item: any;
    }>;
  };

  let readers: Reader[] | undefined;

  return new ReadableStream({
    async pull(controller) {
      try {
        readers ??= streams.filter(isNonNullable).map((s) => ({ reader: s.getReader(), data: [] }));

        while (readers.length) {
          const chunk: Awaited<NonNullable<Reader["reading"]>> = await Promise.race(
            readers.map((i) => {
              i.reading ??= i.reader.read().then((result) => ({ result, item: i }));
              return i.reading;
            }),
          );

          if (chunk.result.value) {
            controller.enqueue(chunk.result.value);
            chunk.item.reading = undefined;
            return;
          }

          if (chunk.result.done) {
            readers = readers.filter((i) => i !== chunk.item);
          }
        }

        controller.close();
      } catch (error) {
        controller.error(error);
        if (readers) for (const item of readers) item.reader.releaseLock();
      }
    },
  });
}
