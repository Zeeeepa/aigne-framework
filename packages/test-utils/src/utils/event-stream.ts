import { readFile } from "node:fs/promises";

export function createMockEventStream<T>(data: { path: string } | { raw: string }): T {
  return new ReadableStream({
    async start(controller) {
      const file = "path" in data ? await readFile(data.path) : data.raw;
      for (const line of file.toString().split("\n")) {
        if (line) controller.enqueue(JSON.parse(line.replace("data:", "")));
      }
      controller.close();
    },
  }) as unknown as T;
}
