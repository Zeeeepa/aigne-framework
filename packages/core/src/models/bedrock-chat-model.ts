import {
  BedrockRuntimeClient,
  type ContentBlock,
  ConverseStreamCommand,
  type ConverseStreamRequest,
  type ConverseStreamResponse,
  type Message,
  type SystemContentBlock,
  type TokenUsage,
  type Tool,
  type ToolChoice,
} from "@aws-sdk/client-bedrock-runtime";
import { nanoid } from "nanoid";
import { z } from "zod";
import type { AgentInvokeOptions, AgentResponse, AgentResponseChunk } from "../agents/agent.js";
import type { Context } from "../aigne/context.js";
import { parseJSON } from "../utils/json-schema.js";
import { getJsonOutputPrompt } from "../utils/prompts.js";
import { agentResponseStreamToObject } from "../utils/stream-utils.js";
import { checkArguments, isNonNullable } from "../utils/type-utils.js";
import {
  ChatModel,
  type ChatModelInput,
  type ChatModelInputMessageContent,
  type ChatModelOptions,
  type ChatModelOutput,
} from "./chat-model.js";

export function extractLastJsonObject(text: string): string | null {
  return text.replace(/<thinking>[\s\S]*?<\/thinking>/g, "").trim();
}

const BEDROCK_DEFAULT_CHAT_MODEL = "us.amazon.nova-lite-v1:0";

export interface BedrockChatModelOptions {
  accessKeyId?: string;
  secretAccessKey?: string;
  region?: string;
  model?: string;
  modelOptions?: ChatModelOptions;
}

export const bedrockChatModelOptionsSchema = z.object({
  region: z.string().optional(),
  model: z.string().optional(),
  modelOptions: z
    .object({
      model: z.string().optional(),
      temperature: z.number().optional(),
      topP: z.number().optional(),
      frequencyPenalty: z.number().optional(),
      presencePenalty: z.number().optional(),
      parallelToolCalls: z.boolean().optional().default(true),
    })
    .optional(),
});

export class BedrockChatModel extends ChatModel {
  constructor(public options?: BedrockChatModelOptions) {
    if (options) checkArguments("BedrockChatModel", bedrockChatModelOptionsSchema, options);
    super();
  }

  protected _client?: BedrockRuntimeClient;

  get client() {
    const credentials =
      this.options?.accessKeyId && this.options?.secretAccessKey
        ? {
            accessKeyId: this.options.accessKeyId,
            secretAccessKey: this.options.secretAccessKey,
          }
        : undefined;
    this._client ??= new BedrockRuntimeClient({
      region: this.options?.region,
      credentials,
    });
    return this._client;
  }

  get modelOptions() {
    return this.options?.modelOptions;
  }

  async process(
    input: ChatModelInput,
    _context: Context,
    options?: AgentInvokeOptions,
  ): Promise<AgentResponse<ChatModelOutput>> {
    const modelId =
      input.modelOptions?.model ?? this.modelOptions?.model ?? BEDROCK_DEFAULT_CHAT_MODEL;

    const { messages, system } = getRunMessages(input);
    const toolConfig = convertTools(input);
    const body: ConverseStreamRequest & { modelId: string } = {
      modelId,
      messages,
      system,
      toolConfig,
      inferenceConfig: {
        temperature: input.modelOptions?.temperature ?? this.modelOptions?.temperature,
        topP: input.modelOptions?.topP ?? this.modelOptions?.topP,
      },
    };

    const command = new ConverseStreamCommand(body);
    const response = await this.client.send(command);
    const jsonMode = input.responseFormat?.type === "json_schema";
    if (options?.streaming && !jsonMode) {
      return this.extractResultFromStream(response.stream, modelId, false, true);
    }

    const result = await this.extractResultFromStream(response.stream, modelId, jsonMode, false);
    return result;
  }

  private async extractResultFromStream(
    stream: ConverseStreamResponse["stream"],
    modelId: string,
    jsonMode: boolean,
    streaming?: false,
  ): Promise<ChatModelOutput>;
  private async extractResultFromStream(
    stream: ConverseStreamResponse["stream"],
    modelId: string,
    jsonMode: boolean,
    streaming: true,
  ): Promise<ReadableStream<AgentResponseChunk<ChatModelOutput>>>;
  private async extractResultFromStream(
    stream: ConverseStreamResponse["stream"],
    modelId: string,
    jsonMode: boolean,
    streaming?: boolean,
  ): Promise<ReadableStream<AgentResponseChunk<ChatModelOutput>> | ChatModelOutput> {
    if (!stream) throw new Error("Unable to get AI model response.");

    const result = new ReadableStream<AgentResponseChunk<ChatModelOutput>>({
      start: async (controller) => {
        try {
          controller.enqueue({ delta: { json: { model: modelId } } });

          const toolCalls: (NonNullable<ChatModelOutput["toolCalls"]>[number] & {
            args: string;
          })[] = [];
          let text = "";
          let usage: TokenUsage | undefined;

          for await (const chunk of stream) {
            if (chunk.contentBlockStart?.start?.toolUse) {
              const toolUse = chunk.contentBlockStart.start.toolUse;
              if (!toolUse.name) throw new Error("Tool use is invalid");
              if (chunk.contentBlockStart.contentBlockIndex === undefined)
                throw new Error("Tool use content block index is required");

              toolCalls[chunk.contentBlockStart.contentBlockIndex] = {
                type: "function",
                id: toolUse.toolUseId || nanoid(),
                function: {
                  name: toolUse.name,
                  arguments: {},
                },
                args: "",
              };
            }
            if (chunk.contentBlockDelta) {
              const block = chunk.contentBlockDelta;
              const delta = block.delta;
              if (delta?.text) {
                text += delta.text;
                if (!jsonMode) {
                  controller.enqueue({ delta: { text: { text: delta.text } } });
                }
              }

              if (delta?.toolUse) {
                if (block.contentBlockIndex === undefined)
                  throw new Error("Content block index is required");
                const call = toolCalls[block.contentBlockIndex];
                if (!call) throw new Error("Tool call not found");
                call.args += delta.toolUse.input;
              }
            }

            if (chunk.metadata) {
              usage = chunk.metadata.usage;
            }
          }

          if (jsonMode && text) {
            const match = extractLastJsonObject(text);
            if (!match) throw new Error("Failed to extract JSON object from model output");

            controller.enqueue({
              delta: { json: { json: parseJSON(match) } },
            });
          }

          if (toolCalls.length) {
            controller.enqueue({
              delta: {
                json: {
                  toolCalls: toolCalls
                    .map(({ args, ...c }) => ({
                      ...c,
                      function: { ...c.function, arguments: parseJSON(args) },
                    }))
                    .filter(isNonNullable),
                },
              },
            });
          }

          controller.enqueue({ delta: { json: { usage } } });
          controller.close();
        } catch (error) {
          controller.error(error);
        }
      },
    });

    return streaming ? result : await agentResponseStreamToObject(result);
  }
}

const getRunMessages = ({
  messages: msgs,
  responseFormat,
}: ChatModelInput): {
  messages: Message[];
  system: SystemContentBlock[];
} => {
  const system: SystemContentBlock[] = [];
  const messages: Message[] = [];

  for (const msg of msgs) {
    if (msg.role === "system") {
      if (typeof msg.content !== "string") throw new Error("System message must have content");
      system.push({ text: msg.content });
    } else if (msg.role === "tool") {
      if (!msg.toolCallId) throw new Error("Tool message must have toolCallId");
      if (typeof msg.content !== "string") throw new Error("Tool message must have string content");
      if (messages.at(-1)?.role === "user") {
        messages.at(-1)?.content?.push({
          toolResult: { toolUseId: msg.toolCallId, content: [{ json: parseJSON(msg.content) }] },
        });
      } else {
        messages.push({
          role: "user",
          content: [
            {
              toolResult: {
                toolUseId: msg.toolCallId,
                content: [{ json: parseJSON(msg.content) }],
              },
            },
          ],
        });
      }
    } else if (msg.role === "user") {
      if (!msg.content) throw new Error("User message must have content");

      messages.push({ role: "user", content: convertContent(msg.content) });
    } else if (msg.role === "agent") {
      if (msg.toolCalls?.length) {
        messages.push({
          role: "assistant",
          content: msg.toolCalls.map(
            (i) =>
              ({
                toolUse: {
                  toolUseId: i.id,
                  name: i.function.name,
                  input: i.function.arguments,
                },
              }) as ContentBlock.ToolUseMember,
          ),
        });
      } else if (msg.content) {
        messages.push({ role: "assistant", content: convertContent(msg.content) });
      } else {
        throw new Error("Agent message must have content or toolCalls");
      }
    }
  }

  if (messages.at(0)?.role !== "user") {
    messages.unshift({ role: "user", content: [{ text: "." }] });
  }

  if (responseFormat?.type === "json_schema") {
    system.push({
      text: getJsonOutputPrompt(responseFormat.jsonSchema.schema),
    });
  }

  return { messages, system };
};

function convertContent(content: ChatModelInputMessageContent): ContentBlock[] {
  if (typeof content === "string") return [{ text: content }];

  if (Array.isArray(content)) {
    const blocks: ContentBlock[] = [];
    for (const item of content) {
      if (item.type === "text") blocks.push({ text: item.text });
    }
    return blocks;
  }

  throw new Error("Invalid chat message content");
}

function convertTools({ tools, toolChoice }: ChatModelInput): ConverseStreamRequest["toolConfig"] {
  if (!tools?.length || toolChoice === "none") return undefined;

  let choice: ToolChoice | undefined;
  if (typeof toolChoice === "object" && "type" in toolChoice && toolChoice.type === "function") {
    choice = { tool: { name: toolChoice.function.name } };
  } else if (toolChoice === "required") {
    choice = { any: {} };
  } else if (toolChoice === "auto") {
    choice = { auto: {} };
  }

  return {
    tools: tools.map((i) => {
      const parameters = i.function.parameters as Record<string, unknown>;
      if (Object.keys(parameters).length === 0) {
        parameters.type = "object";
      }
      return {
        toolSpec: {
          name: i.function.name,
          description: i.function.description,
          inputSchema: { json: parameters },
        },
      } as Tool;
    }),
    toolChoice: choice,
  };
}
