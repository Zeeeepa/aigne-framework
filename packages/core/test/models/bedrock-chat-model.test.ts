import { describe, expect, spyOn, test } from "bun:test";
import { join } from "node:path";
import { BedrockChatModel, extractLastJsonObject } from "@aigne/core/models/bedrock-chat-model.js";
import { parseJSON } from "@aigne/core/utils/json-schema.js";
import { readableStreamToArray } from "@aigne/core/utils/stream-utils.js";
import { createMockEventStream } from "../_utils/event-stream.js";
import {
  COMMON_RESPONSE_FORMAT,
  COMMON_TOOLS,
  createWeatherToolCallMessages,
  createWeatherToolMessages,
} from "../_utils/openai-like-utils.js";

test("BedrockChatModel.invoke 1", async () => {
  const model = new BedrockChatModel();

  spyOn(model.client, "send").mockImplementation(() =>
    Promise.resolve({
      stream: createMockEventStream({
        path: join(import.meta.dirname, "bedrock-streaming-response-1.txt"),
      }),
    }),
  );

  const result = await model.invoke({
    messages: createWeatherToolCallMessages(),
    tools: COMMON_TOOLS,
    responseFormat: COMMON_RESPONSE_FORMAT,
  });

  expect(result).toMatchSnapshot();
});

test("BedrockChatModel.invoke 2", async () => {
  const model = new BedrockChatModel();

  spyOn(model.client, "send").mockImplementation(() =>
    Promise.resolve({
      stream: createMockEventStream({
        path: join(import.meta.dirname, "bedrock-streaming-response-2.txt"),
      }),
    }),
  );
  const result = await model.invoke({
    messages: createWeatherToolMessages(),
    tools: COMMON_TOOLS,
  });

  expect(result).toMatchSnapshot();
});

test("BedrockChatModel.invoke with streaming", async () => {
  const model = new BedrockChatModel();

  spyOn(model.client, "send").mockImplementation(() =>
    Promise.resolve({
      stream: createMockEventStream({
        path: join(import.meta.dirname, "bedrock-streaming-response-text.txt"),
      }),
    }),
  );

  const stream = await model.invoke(
    {
      messages: [{ role: "user", content: "hello" }],
    },
    undefined,
    { streaming: true },
  );

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("BedrockChatModel.invoke without streaming", async () => {
  const model = new BedrockChatModel();

  spyOn(model.client, "send").mockImplementation(() =>
    Promise.resolve({
      stream: createMockEventStream({
        path: join(import.meta.dirname, "bedrock-streaming-response-text.txt"),
      }),
    }),
  );

  const result = await model.invoke({
    messages: [{ role: "user", content: "hello" }],
  });

  expect(result).toMatchSnapshot();
});

describe("extractLastJsonObject", () => {
  test("should extract last JSON object from text", () => {
    const text = `
      <thinking>...</thinking>
      {"key2": "value2", "key3": {"key4": "value4", "key5": [{"key6": "value6"}]}}
    `;
    const match = extractLastJsonObject(text);
    expect(match).toBeTruthy();
    if (match) {
      expect(match).toBe(
        '{"key2": "value2", "key3": {"key4": "value4", "key5": [{"key6": "value6"}]}}',
      );
      expect(parseJSON(match)).toEqual({
        key2: "value2",
        key3: { key4: "value4", key5: [{ key6: "value6" }] },
      });
    }
  });
});
