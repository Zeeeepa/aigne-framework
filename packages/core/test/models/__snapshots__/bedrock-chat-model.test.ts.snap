// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`BedrockChatModel.invoke 1 1`] = `
{
  "json": {
    "text": "The current temperature in New York is 20 degrees Celsius.",
  },
  "model": "us.amazon.nova-lite-v1:0",
  "usage": {
    "inputTokens": 487,
    "outputTokens": 43,
  },
}
`;

exports[`BedrockChatModel.invoke 2 1`] = `
{
  "model": "us.amazon.nova-lite-v1:0",
  "text": 
"<thinking>The User has asked for the current weather in New York. I need to use the 'get_weather' tool to get this information. I will call the tool with the required 'city' parameter.</thinking>
"
,
  "toolCalls": [
    {
      "function": {
        "arguments": {
          "city": "New York",
        },
        "name": "get_weather",
      },
      "id": "tooluse_yPoXvCWrSCScmmoH4UCWOA",
      "type": "function",
    },
  ],
  "usage": {
    "inputTokens": 392,
    "outputTokens": 80,
  },
}
`;

exports[`BedrockChatModel.invoke with streaming 1`] = `
[
  {
    "delta": {
      "json": {
        "model": "us.amazon.nova-lite-v1:0",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "! How can I assist you",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " today? If you have",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " any questions or need information",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " on",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " a particular",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " topic, feel free to let",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " me know. Whether",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " it",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "'s about",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " science, technology",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ", history, or",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " something",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " else, I",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "'m here to help.",
      },
    },
  },
  {
    "delta": {
      "json": {
        "usage": {
          "inputTokens": 1,
          "outputTokens": 51,
          "totalTokens": 52,
        },
      },
    },
  },
  {
    "delta": {
      "json": {
        "model": "us.amazon.nova-lite-v1:0",
        "text": "Hello! How can I assist you today? If you have any questions or need information on a particular topic, feel free to let me know. Whether it's about science, technology, history, or something else, I'm here to help.",
        "usage": {
          "inputTokens": 1,
          "outputTokens": 51,
        },
      },
    },
  },
]
`;

exports[`BedrockChatModel.invoke without streaming 1`] = `
{
  "model": "us.amazon.nova-lite-v1:0",
  "text": "Hello! How can I assist you today? If you have any questions or need information on a particular topic, feel free to let me know. Whether it's about science, technology, history, or something else, I'm here to help.",
  "usage": {
    "inputTokens": 1,
    "outputTokens": 51,
  },
}
`;
